### PR TITLE
Refactor multigrid_ndt_omp

### PR DESCRIPTION
**Pull request by anhnv3991**
Refactored multigrid_ndt_omp.h and multigrid_ndt_omp_impl.hpp to remove unused variables and functions and prepare for the parallel ndt build (which will be added in the future).

### CODE REVIEW

**Comment by SakodaShintaro** [comment]
@anhnv3991 
There seems to be a difference with the previous results, so I would like to find out the cause.
https://github.com/tier4/ndt_omp/actions/runs/8229606913/job/22501142433

If possible, I think it's a good idea to submit only pure refactoring parts where there are no differences in the results as a separate pull request. :seal:

**Comment by anhnv3991** [comment]
@SakodaShintaro I guess the difference is caused by the change from float to double. I'll remove float to double change tomorrow.

File: include/multigrid_pclomp/multigrid_ndt_omp.h [hunk 1]
```diff
@@ -380,9 +380,7 @@ class MultiGridNormalDistributionsTransform : public pcl::Registration<PointSour
    * \param[in] x point from the input cloud
    * \param[in] compute_hessian flag to calculate hessian, unnecessary for step calculation.
    */
-  void computePointDerivatives(Eigen::Vector3d &x, Eigen::Matrix<double, 3, 6> &point_gradient, Eigen::Matrix<double, 18, 6> &point_hessian, bool compute_hessian = true) const;
-
-  void computePointDerivatives(Eigen::Vector3d &x, Eigen::Matrix<float, 4, 6> &point_gradient, Eigen::Matrix<float, 24, 6> &point_hessian, bool compute_hessian = true) const;
```
Comment (SakodaShintaro) [inline include/multigrid_pclomp/multigrid_ndt_omp.h:388]: Do you know why it was 18 for double and 24 for float?

Was `Eigen::Matrix<double, 18, 6> &point_hessian` simply an old implementation that is no longer in use? (It means that the difference is not whether it is a float or a double.)

File: include/multigrid_pclomp/multigrid_ndt_omp_impl.hpp [hunk 2]
```diff
@@ -462,49 +463,51 @@ void pclomp::MultiGridNormalDistributionsTransform<PointSource, PointTarget>::co
...
-  j_ang.row(0).noalias() = Eigen::Vector4f((-sx * sz + cx * sy * cz), (-sx * cz - cx * sy * sz), (-cx * cy), 0.0f);
-  j_ang.row(1).noalias() = Eigen::Vector4f((cx * sz + sx * sy * cz), (cx * cz - sx * sy * sz), (-sx * cy), 0.0f);
-  j_ang.row(2).noalias() = Eigen::Vector4f((-sy * cz), sy * sz, cy, 0.0f);
-  j_ang.row(3).noalias() = Eigen::Vector4f(sx * cy * cz, (-sx * cy * sz), sx * sy, 0.0f);
-  j_ang.row(4).noalias() = Eigen::Vector4f((-cx * cy * cz), cx * cy * sz, (-cx * sy), 0.0f);
-  j_ang.row(5).noalias() = Eigen::Vector4f((-cy * sz), (-cy * cz), 0, 0.0f);
-  j_ang.row(6).noalias() = Eigen::Vector4f((cx * cz - sx * sy * sz), (-cx * sz - sx * sy * cz), 0, 0.0f);
-  j_ang.row(7).noalias() = Eigen::Vector4f((sx * cz + cx * sy * sz), (cx * sy * cz - sx * sz), 0, 0.0f);
+
+  j_ang.row(0) << (-sx * sz + cx * sy * cz), (-sx * cz - cx * sy * sz), (-cx * cy), 0;
```
Comment (SakodaShintaro) [inline include/multigrid_pclomp/multigrid_ndt_omp_impl.hpp]: If the literal is written as '0', won't it be interpreted as an int and a cast from int to double will occur? (Would it be better to write it as '0.0'?) Will it be properly optimized by the compiler?

File: include/multigrid_pclomp/multigrid_ndt_omp_impl.hpp [hunk 3]
```diff
@@ -681,7 +645,7 @@ void pclomp::MultiGridNormalDistributionsTransform<PointSource, PointTarget>::co
 
 //////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 template<typename PointSource, typename PointTarget>
-void pclomp::MultiGridNormalDistributionsTransform<PointSource, PointTarget>::updateHessian(Eigen::Matrix<double, 6, 6> &hessian, const Eigen::Matrix<double, 3, 6> &point_gradient, const Eigen::Matrix<double, 18, 6> &point_hessian, const Eigen::Vector3d &x_trans, const Eigen::Matrix3d &c_inv) const {
+void pclomp::MultiGridNormalDistributionsTransform<PointSource, PointTarget>::updateHessian(Eigen::Matrix<double, 6, 6> &hessian, const Eigen::Matrix<double, 4, 6> &point_gradient, const Eigen::Matrix<double, 24, 6> &point_hessian, const Eigen::Vector3d &x_trans, const Eigen::Matrix3d &c_inv) const {
```
Comment (SakodaShintaro) [inline include/multigrid_pclomp/multigrid_ndt_omp_impl.hpp]: Please let me confirm my understanding.
updateHessian is only used from computeHessian, and computeHessian is only used in the following lines.
https://github.com/tier4/ndt_omp/blob/01ae2426b72ac791bd3695aa9d0660410f1af426/include/multigrid_pclomp/multigrid_ndt_omp_impl.hpp#L937

This is a branch when something is updated in line_search, and since line_search is currently turned off, step_iteration is always 0, so I don't think computeHessian will be called.

Is my understanding correct?
In other words, even if there is a bug in updateHessian, it will not be noticed in its current behavior, and I am concerned that it will occur someday when line_search is turned on.

**Review by SakodaShintaro — COMMENTED** [review]
The difference has become very small. Good

I'm concerned that it's just a little slower, but I think it's OK to merge.

elapsed_milliseconds

I have left some comments and would appreciate it if you could respond.

File: include/multigrid_pclomp/multigrid_ndt_omp.h (hunk 1 above)
Comment (anhnv3991) [reply include/multigrid_pclomp/multigrid_ndt_omp.h:388]: @SakodaShintaro Yes, actually the use of <double, 18, 6> and <double, 3, 6> matrices is from the old PCL implementation, where the columns of point's angular jacobian and hessian (j_ang_ and h_ang_) are defined as individual variables of NDT.
Currently those matrices in PCL are simplified as Eigen::Matrix variables so the computation looks less complicated, but the point gradient and point hessian are still double-type.
Probably the author of ndt_omp found that using float instead of double is faster than using double, while the accuracy is not affected much. So ndt_omp's author decided to use float matrices for point_gradient and point_hessian. As a result, we have a mix-up, where two methods perform the same functionality on different data type.

File: include/multigrid_pclomp/multigrid_ndt_omp_impl.hpp (hunk 3 above)
Comment (anhnv3991) [reply include/multigrid_pclomp/multigrid_ndt_omp_impl.hpp]: @SakodaShintaro If the line search is turned on, errors will 100% occur. It is because the angular derivatives (j_ang_a_, j_ang_b, etc) are not pre-computed. Those variables are used by the computeHessian.

Comment (anhnv3991) [reply include/multigrid_pclomp/multigrid_ndt_omp_impl.hpp:648]: @SakodaShintaro If the line search is turned on, errors will 100% occur. It is because the angular derivatives (j_ang_a_, j_ang_b, etc) are not pre-computed. Those variables are used by the computeHessian.

File: include/multigrid_pclomp/multigrid_ndt_omp_impl.hpp (hunk 2 above)
Comment (anhnv3991) [reply include/multigrid_pclomp/multigrid_ndt_omp_impl.hpp]: @SakodaShintaro I will change they back to 0.0f

File: include/multigrid_pclomp/multigrid_ndt_omp_impl.hpp (hunk 3 above)
Comment (SakodaShintaro) [reply include/multigrid_pclomp/multigrid_ndt_omp_impl.hpp]: @anhnv3991 
I would like to maintain line_search to be able to use.

Steps

(1) Organize parameters: this PR https://github.com/tier4/ndt_omp/pull/43
(2) Add a bool type parameter called use_line_search and make it possible to switch true/false: next PR https://github.com/tier4/ndt_omp/pull/46
(3) I would like to merge this PR while maintaining a buildable state.

**Review by SakodaShintaro — APPROVED** [review]
Looks Good To Me

I apologize for making you wait for such a long time and for causing so many conflicts.

Final result

elapsed_milliseconds

**Comment by anhnv3991** [comment]
@SakodaShintaro Thanks a lot!
